### DISCUSSION TITLE
How to implement functions similar to rostopic [14016]

### DISCUSSION THREAD

**Discussion by zip-link** [eProsima/Fast-DDS#2553] [Q&A] [answered]
In the actual development and debugging, it is often necessary to confirm whether the data published by the publisher is normal when encountering problems. Is there a method or tool that can be implemented? Similar to "rostopic echo" in ROS

**Comment by EduPonz** [comment]
Hi @zip-link 

Ideally, [Fast DDS CLI](https://fast-dds.docs.eprosima.com/en/latest/fastddscli/cli/cli.html) would be [extended](https://github.com/eProsima/Fast-DDS/tree/master/tools/fastdds) to add such functionality, although this is not part of the current roadmap (we are open to contributions though 😉). IMO, the ROS 2 approach is rather static, as the ROS 2 CLI needs to have access to the type libraries for the topics that you want to echo. A more flexible approach for the Fast DDS CLI would be to combine [discovery callbacks](https://fast-dds.docs.eprosima.com/en/latest/fastdds/discovery/disc_callbacks.html) and [dynamic types](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dynamic_types/dynamic_types.html) to be able to create DataReaders on the discovered topics on the fly (without need to know the type description before hand).

One important points is the time for discovery; ROS 2 instantiates a [ROS 2 Daemon](https://github.com/ros2/ros2cli/blob/master/ros2cli/ros2cli/node/daemon.py) with the very first CLI command you run, and this daemon maintains a discovery graph which is made accessible to the `ros2 topic list|echo` commands. That way, the `list` command does not need to wait a somewhat arbitrary time to consider discovery "finished". I myself am not particularly fond of installing and running  a background application which the user does not instantiate directly and that is stopped after an also arbitrary time; in fact, my experience is that most people do not even know that is there! Instead, I'd have a `fastdds discovery show` command running either for a configurable time or until the user closes it, showing a table which updates with every discovered entity (topics, participants, datareaders, datawriters).

**Reply by zip-link** [reply]
As far as I know, fast DDS monitor can already view the relationship between various participants and topics. If this tool extends the serialization function, can it realize the above functions?

**Comment by jparisu** [comment, marked as answer]
Hi @zip-link 

## Fast DDS Monitor
This is an idea that we already considered. It would be awesome to have such functionality in Fast DDS Monitor.
However, the time and effort needed would be high, and we do not plan to do it so far.

## DDS Router 
- https://github.com/eProsima/DDS-Router

We have implemented a new DDS Router tool. Its main application is to communicate DDS networks in WAN.
However, it has a functionality that may be of your interest.
There is an `echo` Participant that prints the data arrived (in hexadecimal). Check this example on how it works: https://eprosima-dds-router.readthedocs.io/en/latest/rst/examples/echo_example.html . The good part is that it is independent from the topic and/or type (it could be agnostic to data type) and can be filtered in execution time to see one topic or another.
This feature is mostly used by our team to debug communications, but you may see it useful regarding your use case.
